eland/ml/pytorch/pytorch_model.py

```diff
@@ -0,0 +1,154 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
+#  license agreements. See the NOTICE file distributed with
+#  this work for additional information regarding copyright
+#  ownership. Elasticsearch B.V. licenses this file to you under
+#  the Apache License, Version 2.0 (the "License"); you may
+#  not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+import base64
+import json
+import math
+import os
+from typing import TYPE_CHECKING, Any, Dict, List, Set, Tuple, Union
+
+import elasticsearch
+from tqdm.auto import tqdm
+
+from eland.common import ensure_es_client
+
+if TYPE_CHECKING:
+    from elasticsearch import Elasticsearch
+
+DEFAULT_CHUNK_SIZE = 4 * 1024 * 1024  # 4MB
+QUEUE_SIZE = 4
+THREAD_COUNT = 4
+DEFAULT_TIMEOUT = "60s"
+
+
+class PyTorchModel:
+    """
+    A PyTorch model managed by Elasticsearch.
+
+    These models must be trained outside of Elasticsearch, conform to the
+    support tokenization and inference interfaces, and exported as their
+    TorchScript representations.
+    """
+
+    def __init__(
+        self,
+        es_client: Union[str, List[str], Tuple[str, ...], "Elasticsearch"],
+        model_id: str,
+    ):
+        self._client = ensure_es_client(es_client)
+        self.model_id = model_id
+
+    @staticmethod
+    def _load_json(path: str) -> Dict[str, Any]:
+        with open(path, "r") as infile:
+            return json.load(infile)
+
+    def _upload_config(self, path: str) -> bool:
+        config = PyTorchModel._load_json(path)
+        response = self._client.ml.put_trained_model(
+            model_id=self.model_id, body=config
+        )
+        return response is not None
+
+    def _upload_vocab(self, path: str) -> bool:
+        vocab = PyTorchModel._load_json(path)
+        return self._client.transport.perform_request(
+            method="PUT",
+            url=f"/_ml/trained_models/{self.model_id}/vocabulary",
+            body=vocab,
+        )
+
+    def _upload_model(
+        self, model_path: str, chunk_size: int = DEFAULT_CHUNK_SIZE
+    ) -> bool:
+        file_stats = os.stat(model_path)
```

Comment:
nit: Since `file_stats.st_size` appears to be the only thing used perhaps we can save that value instead of the whole stats object?